### PR TITLE
Core: Declarative story definition in main.js

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: preview.js is required to inject stories, this is in imperative code, and impossible to extract. If we declaratively configure the stories in main.js, we could determine the stories without running storybook.

## What I did
Make it possible to define stories entries in main.js

Hopefully with enough minor steps like this we can make it less and less likely users need preview.js ever.

### CODE REVIEW

**Comment by tmeasday** [comment]
@ndelangen this is awesome!

I'm not sure if it should be in this PR or seperately, but we need to update storyshots to support this. It'll be great though. @shilman and I discussed it a bit, and here's what should happen:

Right now to use SS:

1. You write a `.test.js` file that imports and calls `initStoryshots`
2. `initStoryshots` ends up `require`-ing your `.storybook/config.js`
3. You `.storybook/config.js`, which itself:
  a) does whatever configuration (adding decorators etc)
  b) then calls [`requireContext()`](https://www.npmjs.com/package/require-context.macro) a `require.context()` macro that in Jest actually just ends up running a bunch of `require()` calls.

This will actually simplify things and remove the need for that macro which would be awesome! To support this properly, we'd need to:

1. Support triconfig properly in storyshots. I think this probably just means also requiring `.storybook/preview.js` in step 2. above.

2. After doing step 2-3 above, `initStoryshots` should then: require `.storybook/main.js`, pull the `stories` key out, then require all the stories.

**Comment by ndelangen** [comment]
@tmeasday I propose to tackle that in another PR.

If we release this without storyshots support, users can still use the 'old way of doing things'

**Comment by ndelangen** [comment]
@tmeasday https://github.com/storybookjs/storybook/pull/8765

**Review by shilman — APPROVED** [review]
This looks fantastic @ndelangen 👏👏👏 

Would love to go through it together when you're available and also get feedback from others, since it's a major change

**Review by shilman — APPROVED** [review]
This looks fantastic @ndelangen 👏👏👏 

Would love to go through it together when you're available and also get feedback from others, since it's a major change

File: lib/core/src/server/preview/iframe-webpack.config.js [hunk 1]
```diff
@@ -44,6 +71,16 @@ export default ({
       publicPath: '',
     },
     plugins: [
+      stories && stories.length
+        ? new VirtualModulePlugin({
+            [path.resolve(path.join(configDir, `generated-entry.js`))]: `
+              import { configure, addDecorator, addParameters } from '@storybook/react';
```
Comment (shilman) [inline lib/core/src/server/preview/iframe-webpack.config.js]: Does this need to change to `@storybook/${framework}`??

**Review by shilman — CHANGES_REQUESTED** [review]
Seems to be react-specific at this point

File: lib/core/src/server/preview/iframe-webpack.config.js (hunk 1 above)
Comment (shilman) [reply lib/core/src/server/preview/iframe-webpack.config.js]: Framework is currently used in all the presets from `options.framework` 😉

**Comment by shilman** [comment]
Also can you upgrade one more non-react example to test?

File: lib/core/src/server/preview/iframe-webpack.config.js (hunk 1 above)
Comment (ndelangen) [reply lib/core/src/server/preview/iframe-webpack.config.js]: yes

Comment (ndelangen) [reply lib/core/src/server/preview/iframe-webpack.config.js]: ✅

**Comment by ndelangen** [comment]
> Also can you upgrade one more non-react example to test?

I can't without adding storyshots support

I have migrated the html example in this PR https://github.com/storybookjs/storybook/pull/8765

**Comment by ndelangen** [comment]
@shilman is this all good to go?

**Comment by ndelangen** [comment]
EPIC WIN

**Pull request merged**